scalafmt-sbt/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala

```diff
@@ -1,25 +1,175 @@
 package org.scalafmt.sbt
+
+import org.scalafmt.config.{Config, ScalafmtConfig}
+import org.scalafmt.{Formatted, Scalafmt}
+import sbt.Keys._
+import sbt.{Def, _}
+import complete.DefaultParsers._
+import metaconfig.Configured
+import sbt.util.Logger
+
 import scala.meta.internal.tokenizers.PlatformTokenizerCache
-import sbt._
-import Keys._
+import scala.util.{Failure, Success, Try}
 
 object ScalafmtPlugin extends AutoPlugin {
   override def trigger: PluginTrigger = allRequirements
+
   object autoImport {
-    val scalafmt: Command =
-      Command.args("scalafmt", "run the scalafmt command line interface.") {
-        case (state, args) =>
-          org.scalafmt.cli.Cli.main("--non-interactive" +: args.toArray)
-          PlatformTokenizerCache.megaCache.clear()
-          state
+    val scalafmt = taskKey[Unit]("Format Scala sources")
+    val scalafmtCheck =
+      taskKey[Boolean]("Check that Scala sources is formatted properly")
+    val scalafmtOnCompile =
+      settingKey[Boolean]("Format source when compiling")
+    val scalafmtConfig = taskKey[File]("Scalafmt config file")
+    val scalafmtSbt = taskKey[Unit]("Format SBT sources")
+    val scalafmtSbtCheck =
+      taskKey[Boolean]("Check that SBT sources is formatted properly")
+    val scalafmtOnly = inputKey[Unit]("Format only given file")
+  }
+  import autoImport._
+
+  private val scalaConfig = scalafmtConfig.map(Config.fromHoconFile(_) match {
+    case Configured.Ok(conf) => conf
+    case Configured.NotOk(e) => throw new MessageOnlyException(e.msg)
+  })
+  private val sbtConfig = scalaConfig.map(_.forSbt)
+
+  private type Input = String
+  private type Output = String
+
+  private def withFormattedSources[T](
+      sources: Seq[File],
+      config: ScalafmtConfig
+  )(
+      onError: (File, Throwable) => T,
+      onFormat: (File, Input, Output) => T
+  ): Seq[Option[T]] = {
+    sources.map(
+      file => {
+        val input = IO.read(file)
```

Comment:
Would it make sense to make this a cached function? http://www.scala-sbt.org/1.x/docs/Faq.html#How+can+a+task+avoid+redoing+work+if+the+input+files+are+unchanged%3F

Scalafmt is around 6x slower than scalariform because it does a lot of work to figure out the best line wrapping (and the best-first algorithm I used is quite a bad fit for this domain, see https://github.com/scalameta/scalafmt/issues/917). Scalafmt performance is usually not a problem for smaller files or if you run scalafmt infrequently but if you run it on compile on large source files it can impact compile times. I'll leave it to you to decide, I would at least keep it in mind.